### PR TITLE
Add minification process

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
r? @QuietMisdreavus

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
For the record, here's the error:

```
Building rustdoc for stage1 (x86_64-apple-darwin)
   Compiling pulldown-cmark v0.1.2
   Compiling tempdir v0.3.7
   Compiling minifier v0.0.4
   Compiling rustdoc v0.0.0 (file:///Users/imperio/rust/rust/src/librustdoc)
   Compiling rustdoc-tool v0.0.0 (file:///Users/imperio/rust/rust/src/tools/rustdoc)
    Finished release [optimized] target(s) in 54.79 secs
dyld: Library not loaded: @rpath/libminifier-ff582a1ef3309921.dylib
  Referenced from: /Users/imperio/rust/rust/build/x86_64-apple-darwin/stage1/bin/rustdoc
  Reason: image not found
```

**Comment by ollie27** [comment]
Is there a reason you are tying to build it as a dylib? https://github.com/GuillaumeGomez/minifier-rs/blob/81851f08301dd149c2de25889ade81291ca6de74/Cargo.toml#L23

**Comment by GuillaumeGomez** [comment]
Old (and bad apparently) habit. Good catch, thanks a lot!

**Comment by GuillaumeGomez** [comment]
Fixed the initial issue, now just need to make the minification process work as expected. :)

File: src/librustdoc/html/render.rs [hunk 1]
```diff
@@ -814,16 +818,20 @@ themePicker.onclick = function() {{
...
 
-    write(cx.dst.join(&format!("main{}.js", cx.shared.resource_suffix)),
-                      include_bytes!("static/main.js"))?;
-    write(cx.dst.join(&format!("settings{}.js", cx.shared.resource_suffix)),
-                      include_bytes!("static/settings.js"))?;
+    write_minify(cx.dst.join(&format!("main{}.js", cx.shared.resource_suffix)),
+                 include_bytes!("static/main.js"),
+                 enable_minification)?;
+    write_minify(cx.dst.join(&format!("settings{}.js", cx.shared.resource_suffix)),
+                 include_bytes!("static/settings.js"),
```
Comment (kennytm) [inline src/librustdoc/html/render.rs]: Could we minimize these two `*.js` files at `build.rs` of rustdoc?

Comment (GuillaumeGomez) [reply src/librustdoc/html/render.rs]: I'd rather prefer that we don't. We *could* but then disabling minification would be a bit more complicated and I'm not sure it's worth it.

File: src/librustdoc/html/render.rs [hunk 2]
```diff
@@ -1020,6 +1028,18 @@ fn write(dst: PathBuf, contents: &[u8]) -> Result<(), Error> {
     Ok(try_err!(fs::write(&dst, contents), &dst))
 }
 
+fn write_minify(dst: PathBuf, contents: &[u8], enable_minification: bool) -> Result<(), Error> {
+    if enable_minification {
+        if let Ok(s) = String::from_utf8(contents.to_vec()) {
```
Comment (ollie27) [inline src/librustdoc/html/render.rs]: Use `include_str!` instead of `include_bytes!` for the static files to avoid this UTF-8 check.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render.rs]: Thanks!

**Comment by GuillaumeGomez** [comment]
Updated following @ollie27's comment.

**Comment by ollie27** [comment]
Yeah this looks good.

For the `std` docs it might make more sense for the minification to be done by rustbuild so it can apply to the docs generated by mdBook as well but this seems like a reasonable start anyway.

**Comment by GuillaumeGomez** [comment]
For now it's experimental. If we have no issue for a given time, we might want to expand this to other tools as well. Thanks for your review!

@bors: r=ollie27

**Pull request merged**